doc/api-details.md

```diff
@@ -164,6 +176,61 @@ $ curl -X 'PUT' \
 {"_id":"61bda8f2eb1a63d2b7152418","kind":"node","name":"checkout-test","revision":{"tree":"mainline","url":"https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git","branch":"master","commit":"2a987e65025e2b79c6d453b78cb5985ac6e5eb26","describe":"v5.16-rc4-31-g2a987e65025e"},"parent":null,"status":"pending","result":null, "created":"2022-02-02T11:23:03.157648", "updated":"2022-02-02T12:23:03.157648"}
 ```
 
+### State diagram
+
+The Node objects are governed by the following state machine:
+
+```mermaid
+graph LR
+  running(Running) --> job{successful?}
+  running -.-> |timeout| done(Done)
+  job --> |no| done(Done)
+  job --> |yes| ready(Ready)
+  ready -.-> |timeout| done
+  ready --> |holdoff expired| pending(Pending)
+  pending -.-> |timeout| done
+  pending --> |child nodes done| done
+```
+
+The state of the Node is kept in the `state` field in the `Node` models.  The
+different state values are described below:
+
+Running
+: The job has been scheduled and there is no result yet.
+
+Ready
+: The job was successful.  Child nodes that depend on the job's success can now
+  be created.
+
+Pending
```

Comment:
How about renaming it to "awaiting" as we are waiting for the child nodes to get completed?
```suggestion
Awaiting
```

Reply:
I'm not sure, because it could be "awaiting job completion" or anything.  It doesn't really mean it's awaiting child nodes to complete.

I think it's better if the names are about the node state seen from the node's "point of view".  So "running" and "done" are fine, but "ready" means it's waiting for any potential child nodes to get scheduled.  The node itself is not really more ready when it's in this state than when it's "running".  The "pending" or "awaiting" state has the same issue I think.

So here are the definitions of these states, let's call them 2 and 3 (1 being running and 4 being done):

2: The job has succeeded, any other job that depend on it can now get scheduled (e.g. tests depending on a source tarball).  The "holdoff" time has not been reached yet, so the node won't move to the Done state even if all child nodes are already Done.

3: The "holdoff" time has been reached, so as soon as all child nodes are Done the node will also be moved to the Done state.

Reply:
I guess state 2 could be called Available (instead of Ready) even though that's arguably not really from the node's point of view, it's available for other jobs to use it.

For state 3, maybe Closing would be a better fit.  Like, it's time to get out of a shop or something before it's too late (timeout).  And that's really about the parent node, it's closing so child nodes have to complete in time.

A more poetic way to call these states could be Dawn, Day, Dusk and Night.  I think we could use that for a presentation slide to illustrate the states rather than actually call them like that in the API as it's probably a bit too vague.

Reply:
When I think from the node's point of view this makes more sense to me:
```
Running --> Available --> Closing/Completing/Terminating/Ending --> Done
```

2. `Available` is preferable here as it denotes that a node is available for creating child nodes.
3. Here we are trying to say is the node is about to complete or end as the hold off has been reached. I would rather go for one of `Closing/Completing/Terminating/Ending` states to denote it. The `pending` will not convey the actual point.

Reply:
OK thanks.  So I think we agree on Running, Available, Closing, Done